### PR TITLE
[3.x] Fix multiple issues with test_body_motion() and test_body_ray_separaton().

### PR DESCRIPTION
**Pull request by madmiraal**
This is a the 3.2 branch version of PR #35945.

I've created this because #34098, #34436, #34596, #35713 and #35780 are regression issues from 3.1 and since master has changed so much, #35945 is no longer cherry-pickable.

Update:
Now includes additional fixes added to #35945 for #17893, #18433, #26550, #28717 and #32182.

### CODE REVIEW

**Comment by madmiraal** [comment]
As with #35945, updated with fixes to #18433, #26550 and #32182 and improvement to #31981.

**Comment by travisryte** [comment]
I tested this and it works on the recent 3.2

**Comment by travisryte** [comment]
test
Tested it with his build, it works apparently

**Comment by MBZSoftProds** [comment]
Hi, travisryte shared his compiled version and I've been testing it with my project. With no code changes to my character controller, the bumps which happen when crossing from one collider to another on the floor are gone. However, the body occasionally gets stuck if I'm touching both the floor and the wall at the same time. is_on_floor() becomes unstable then.

**Comment by Zireael07** [comment]
IIRC is_on_floor being unstable when you touch two things is a different preexisting problem?

**Comment by MBZSoftProds** [comment]
Here's a video of what I mean.
[Link](https://streamable.com/uqmb6z)

I'm simply doing this in my code:

`velocity = move_and_slide_with_snap(velocity, Vector3.DOWN, Vector3.UP)  `

**Comment by MBZSoftProds** [comment]
> 
> 
> IIRC is_on_floor being unstable when you touch two things is a different preexisting problem?

No, it's new with this fix.

**Comment by madmiraal** [comment]
@MBZSoftProds, please share your project; so I can investigate.

**Comment by MBZSoftProds** [comment]
Here you go.

[Link](https://easyupload.io/xekhxr)

Hope this works. It's rather large.

**Comment by jknightdoeswork** [comment]
I had a regression in my golf game where the ball's y position would go up and down when on a seam. This PR led me to find that its a regression in Godot 3.2.

Looks like I'm downgrading to 3.1.

In 3.2, if the ball comes to rest on a seam, it resolves the collision _slightly into the air_ then is no longer colliding with anything, and drops back down to the ground.

This also caused my golf ball to have a different speed, as it has different friction in the air and on the ground, so a whole host of problems were introduced due to 3.2.

I came up with a complicated hack to prevent the ball from depentrating into the air when on a seam. I just tested in 3.1 and sure enough, its not there.

Thanks for at least alerting me to this regression.

**Comment by madmiraal** [comment]
Added a fix for the sticking identified by @MBZSoftProds.

Note: There is still some stickiness, and some `on_floor()` instability, but I agree with @Zireael07 this was there before, appears to be due to something else, and may be related to #29392.

**Comment by asheraryam** [comment]
Would be really nice to have this merged for the upcoming release, let me know if there is anything I can do to help with that

**Comment by DenisBelmondo** [comment]
Hi @madmiraal, great fix so far. I've noticed some issues with regards to KinematicBody sticking. Here are some of my observations:

Using the stock `move_and_slide` and `move_and_slide_with_snap` (both are affected by this issue, ruling out the possibility that it is `_with_snap`'s mechanisms), a KinematicBody occasionally sticks to surfaces. This includes floors and walls. With floors, you can eliminate the issue entirely by doing a floor check with `test_move(relVec=Vector3.DOWN)` or `move_and_collide(test_only=true)` and setting the y velocity to 0, but the same approach cannot be used for walls.

Below is a gif of the player (using a BoxShape, but still reproducible with a CapsuleShape in my experience) attempting to slide across a continuous wall (_not_ made of GridMap tiles or multiple meshes) but getting stuck along the way.

stuck1

When you remove the floor check however, you can see that the character gets stuck in it intermittently:

stuck2

In my experience, setting the CollisionShape margin within reasonable bounds (0.04 to 0.2) did not have an effect. Anything beyond those margin values seemed to either worsen it or have no effect at all.

[stucktest.zip](https://github.com/godotengine/godot/files/4565783/stucktest.zip)

Here is a minimal reproduction project. This was created with the [fix-34596-3.2](https://github.com/madmiraal/godot/tree/fix-34596-3.2) branch, **compiled with Mono.** Apologies if this is an inconvenience, as my project is being written in C#.

To test just the walls, insert
```
if (TestFloor())
{
    Vel.y = 0f;
}
```

at the beginning of `_PhysicsProcess`.

**Comment by MBZSoftProds** [comment]
Tried the fix-34596-3.2 branch too. It works much better than before. It fixes the floor issues still and wall sticking is almost gone. It only happens rarely and to a much lesser extent than before. 

However, multiple moving KinematicBody entities, i.e. AI controlled entities, together with the player introduces regular slow downs and stutter for around a second ...

**Comment by jknightdoeswork** [comment]
Tested this in my golf game. When the golf ball is on the ground and gravity is present, move_and_collide will no longer apply horizontal forces. Seems like the gravity causes the golf ball to instantly collide and instantly stop, the perpendicular velocity vector is not applied to the transform at all. Is this intended behaviour?

**Comment by madmiraal** [comment]
@DenisBelmondo This is the residual stickiness (and floor instability) I'm referring [to](https://github.com/godotengine/godot/pull/37498#issuecomment-619060262). As @jknightdoeswork pointed out, and beautifully illustrated by @Zylann [here](https://github.com/godotengine/godot/issues/28032#issuecomment-504799075), this appears to be due to the effective rounding of box shapes created by the collision margin and the non-perpendicular collision normals returned when colliding with the edge of a `BoxShape`.

This non-perpendicular normal will push the `KinematicBody` back (or to the side as in #29392) when using `move_and_slide()` or `move_and_slide_with_snap()`. This happens even when colliding with two aligned `BoxShapes`, because the collision with each `CollisionShape` is calculated independently and then collated. If the motion towards two aligned `BoxShapes` is greater than the motion along them, the `KinematicBody` gets stuck, and if the collated normal is more than 45° off the vertical, the `on_floor()` returns `false`. This is why setting the `y` velocity to zero works: there is no motion towards the boxes.

In short, this is a separate, pre-existing issue, and addressing it is not part of this PR.

**Comment by madmiraal** [comment]
@MBZSoftProds I think the stuttering you're seeing may be related to #37118.

**Comment by madmiraal** [comment]
@jknightdoeswork I'm confused by what you're asking. `KinematicBodies` don't have forces applied to them and [`move_and_collide()`](https://docs.godotengine.org/en/stable/classes/class_kinematicbody.html#class-kinematicbody-method-move-and-collide) doesn't return an updated `linear_velocity` vector.

If you're referring to the sliding seen in #18433, this was a bug, which has been resolved with this PR.

**Comment by jknightdoeswork** [comment]
@madmiraal 

When a sphere is on a box, calling `move_and_collide` on the sphere with a velocity of (-10.0, -1.0, 0) will not move the sphere. If you disable gravity and apply (-100, 0, 0). The sphere moves as expected.

Before this patch, it would.

In my golf game, I am integrating forces myself and using move_and_collide in order to move them. I do this so I can have very custom control over the bounce off the wall.

Before the patch, when the ball was on the ground, it would behave much differently.
After this patch, when the ball is on the ground it cannot be moved horizontally.

I can make a reproduction project if needed.

**Comment by jknightdoeswork** [comment]
@DenisBelmondo You can try to get exact normals by using Concave shape for your walls/ground. You can do this by creating a cube mesh, then clicking Mesh -> Create Trimesh Static Body (above the viewport), and then using that collision shape.

**Comment by asheraryam** [comment]
I am working on an identical project as @jknightdoeswork -- a minigolf game where I use move_and_collide to handle bouncing and movement, with collisions happening between a sphere and Trimesh Static Bodies.

I can try to provide a stripped down project if it is necessary for comparison also.

@jknightdoeswork We should get in touch and share our experiences :b

**Comment by jknightdoeswork** [comment]
@asheraryam Definetly should. I'm 00jknight in the godot discord.

@madmiraal This code block causes the patch to behave similarly to before the patch:
`
var collision = move_and_collide(v, true, true, false)
if collision != null:
	var projected_remainder = collision.remainder.project(collision.normal)
	translation += collision.remainder - projected_remainder
`
Again, I'm not sure if this is better or worse than before the patch. At first glance, it seems like a regression, but upon reflection you can see that the patch is now causing move_and_collide to behave more like how it claims to behave in it's documentation (instantly stopping on a collision).

Note that this branch resolves the "seam bouncing" issue that I was previously seeing. So great job on that!

**Comment by DenisBelmondo** [comment]
Hi @madmiraal, apologies, I misunderstood the scope of this PR. 😅

For what it's worth though, I can further corroborate once more that the worst jitter-causing entity since even 3.1 (any trimesh collision shape) is, according to my testing, completely gone.

Unfortunately, @jknightdoeswork, the KinematicBody player still sticks to a singular concave wall in my experience (arranged contiguously with others or not). Since I now understand this to be irrelevant to this PR, I'll continue the discussion elsewhere, but I thank you for the recommendation.

**Comment by jknightdoeswork** [comment]
@DenisBelmondo Are you using `move_and_slide` or `move_and_collide`?

If your using `move_and_collide`, you could try this code block:

```
    var collision = move_and_collide(v, true, true, false)
    if collision != null:
        var projected_remainder = collision.remainder.project(collision.normal)
        translation += collision.remainder - projected_remainder
```

If your using `move_and_slide`, it could be a problem in the implementation of `move_and_slide`?

**Comment by hoontee** [comment]
Would turning this into a project setting make it viable for a 3.2.2/3 release?

**Comment by jknightdoeswork** [comment]
Just to clarify, I am using this patch in IsoPutt and plan on releasing it
using this as it fixes some other problems that I was having with mesh
seams. It does drastically change `move_and_collide`, but overall I support
it.

It's possible that we'd want to apply the remainder to the translation
inside `move_and_collide` though. (See my previous comment for the
project/add code)

Additionally we should really put `move_and_slide` through the rigor of
testing because we don't want to introduce more regressions.

We should probably have a standard test scene to test character controllers
for this kind of stuff.

**Comment by evsey9** [comment]
Any clue when this is getting merged? 3.2.2 released, and the issues that this PR fixes still persist in it.

**Comment by jknightdoeswork** [comment]
I think we could use a review from @AndreaCatania  here as he wrote a lot of this code.

**Comment by madmiraal** [comment]
Rebased following #39084.

File: modules/bullet/space_bullet.cpp [hunk 1]
```diff
@@ -890,202 +890,150 @@ void SpaceBullet::update_gravity() {
...
+					} else {
+						safe = mid_point;
+					}
+				}
+				motion *= safe;
 			}
 		}
+
+		B_TO_G(motion + recover_motion, r_motion_result->motion);
+		B_TO_G(motion, r_motion_result->remainder);
```
Comment (jknightdoeswork) [inline modules/bullet/space_bullet.cpp:1015]: This line has no effect and should be removed.

File: modules/bullet/space_bullet.cpp [hunk 2]
```diff
@@ -890,202 +890,150 @@ void SpaceBullet::update_gravity() {
...
-				r_result->remainder = Vector3();
+				B_TO_G(collision_result.pointWorld, r_motion_result->collision_point);
+				B_TO_G(collision_result.normal, r_motion_result->collision_normal);
+				B_TO_G(btRigid->getVelocityInLocalPoint(collision_result.pointWorld - btRigid->getWorldTransform().getOrigin()), r_motion_result->collider_velocity);
+				r_motion_result->collider = collisionObject->get_self();
+				r_motion_result->collider_id = collisionObject->get_instance_id();
+				r_motion_result->collider_shape = collision_result.other_compound_shape_index;
+				r_motion_result->collision_local_shape = collision_result.local_shape_most_recovered;
+
+				// Use binary search to find the safe motion.
```
Comment (jknightdoeswork) [inline modules/bullet/space_bullet.cpp:996]: This section seems strange. Why is a binary search used to find the safe motion? Shouldn't btResult.m_closestHitFraction already contain the safe motion? And is that value not already applied to motion from this code block:

```
// Adjust the motion so only closer shapes will also hit.
motion *= btResult.m_closestHitFraction;
```

Comment (jknightdoeswork) [reply modules/bullet/space_bullet.cpp:996]: I dont understand why there is a "step 3" that has to call "kinematic_collision_test" again.

Why is the algorithm incomplete after performing `dynamicsWorld->convexSweepTest`?

File: modules/bullet/space_bullet.cpp (hunk 1 above)
Comment (madmiraal) [reply modules/bullet/space_bullet.cpp:1015]: Good spot. This is a remnant of the old code. I'll remove it.

File: modules/bullet/space_bullet.cpp (hunk 2 above)
Comment (madmiraal) [reply modules/bullet/space_bullet.cpp:996]: The `btResult.m_closestHitFraction` returns the proportion of the move that will result in a collision, we need to find the proportion of the move that will **not** result in a collision: the safe distance. Furthermore, the `convexSweepTest` doesn't detect all collisions; so sometimes `btResult.m_closestHitFraction` is 1.0. A binary search is an effective means to find this proportion to any level of accuracy. With 8 steps (used in Godot physics too) the precision is 0.39% of the total move distance.

**Comment by madmiraal** [comment]
Removed superfluous line as suggested.

File: modules/bullet/space_bullet.cpp (hunk 2 above)
Comment (jknightdoeswork) [reply modules/bullet/space_bullet.cpp:996]: > The btResult.m_closestHitFraction returns the proportion of the move that will result in a collision, we need to find the proportion of the move that will not result in a collision

Could this be `btResult.m_closestHitFraction - 0.001` ?

> Furthermore, the convexSweepTest doesn't detect all collisions

Which collisions are not detected by convex sweep test? Concave shapes?

Comment (madmiraal) [reply modules/bullet/space_bullet.cpp:996]: > Could this be btResult.m_closestHitFraction - 0.001 ?

This would only work if we knew that `m_closestHitFraction` had that level of accuracy, which we don't; if only, because it sometimes returns 1.0.

> Which collisions are not detected by convex sweep test? Concave shapes?

I haven't done a thorough analysis, but it was more than just concave shapes, which is why I included the binary search.

**Comment by madmiraal** [comment]
I've rebased this PR following cherry-picking of #40377; so people using this branch can have its benefit too.

**Comment by madmiraal** [comment]
Rebased following #40490.

**Comment by janmagtoast** [comment]
I've really been enjoying using this fix (especially when trying to write a move_and_slide myself that works the way I want it to work).

But I've been having a huge issue. From what I can tell, it breaks area_body_entered() and area_area_entered().
I've tested it in a very simple scene with only one kinematic body moving into an area and a print function to tell me whether it entered the area. 
I've tested it in 3.2 and in 3.2 with all the files changes in this version.
3.2 works just fine but sadly this version isn't calling area_body_entered().

**Comment by madmiraal** [comment]
@janmagtoast This fix doesn't change how `Areas` work; so I'd be very surprised if the issue you're experiencing is caused by this PR. I assume you're talking about `Area` signals. I've tested `Area` signals with this PR and they work as expected. Please share a minimal reproduction project to demonstrate the issue you're experiencing?

**Comment by janmagtoast** [comment]
@madmiraal I don't know what happened last time, but I recompiled the engine and now it's working as intended.

**Comment by stoikerty** [comment]
I would like to use Gridmap without the jitter issues and it seems that this PR is the one that fixes it on the current 3.2 version. I've been waiting patiently for weeks for the fix to go out but after seeing that it didn't get released with 3.2.2 I'm now trying to compile godot from source although I have to admit I'm struggling a bit 😢 

After cloning the project I am unable to switch to this branch with `git checkout -b madmiraal:fix-34596-3.2`, this project seems to use a branch tracking method I'm not familiar with. How do I go about doing this?

Alternatively, does someone have an already compiled version of this fix (for windows) that I could download? That would be a massive help!

**Comment by stoikerty** [comment]
Ah nevermind, I got it. I just realised this is a PR from an external contribution so I had to do the following...
```sh
git clone git@github.com:madmiraal/godot.git madmiraal-godot
git checkout fix-34596-3.2
```

I'm now compiling it by using the Scoop & Msys2 method [mentioned in the docs](https://docs.godotengine.org/en/stable/development/compiling/compiling_for_windows.html#compiling-for-windows) (avoiding the multi-gig install that is visual studio community!)
```sh
# installed https://scoop.sh/ and ran in powershell
scoop install gcc python scons yasm make

# installed https://www.msys2.org/ and inside the msys window I ran
pacman -S mingw-w64-x86_64-python3-pip \
    mingw-w64-x86_64-gcc mingw-w64-x86_64-yasm \
    mingw-w64-i686-python3-pip mingw-w64-i686-gcc \
    mingw-w64-i686-yasm make

# then opened powershell and ran the following to compile
scons platform=windows
```

still waiting for it to finish but hoping this will work 🤞

**Comment by madmiraal** [comment]
@stoikerty As mentioned [above](https://github.com/godotengine/godot/pull/37498#issuecomment-619060262), the floor stickiness you're seeing is a separate issue related to #29392.

**Comment by vesk4000** [comment]
Great commit can it be included in stable?

**Review by vesk4000 — APPROVED** [review]
Great commit, very important changes that need to be included in Godot as soon as possible.

**Comment by madmiraal** [comment]
Rebased following merge of #42575.

**Comment by Two-Tone** [comment]
Is there any chance of this being merged soon?

**Comment by vesk4000** [comment]
> Is there any chance of this being merged soon?

To be honest, even though I like the PR, I doubt it. On the other hand a lot of the physics issues seem to be getting fixed for 4.0 thanks to @pouleyKetchoupp

**Comment by Two-Tone** [comment]
It being fixed in 4.0 is great and all, but that is irrelevant for people
that expect to be using 3.2 long term.

**Comment by Riteo** [comment]
> Is there any chance of this being merged soon?

I highly doubt it. If you look at the 4.0 version of this PR you'll see how this set of changed isn't planned to be merged, but instead is kept open just to be able to salvage whatever is considered good by @pouleyKetchoupp.

In any case though, you can just merge this PR in your own fork and call it a day.

**Comment by Two-Tone** [comment]
And again, what about all the other people who'll be still using 3.2? Do you really expect thousands of developers to support a custom fork of 3.2 just for fixes like these when they could have just been merged instead of going "It's fixed in 4.0, fuck em"?

The whole "if you don't like it make your own fork" attitude is unnecessarily hostile towards users and gamedevs.

**Comment by Riteo** [comment]
> And again, what about all the other people who'll be still using 3.2? Do you really expect thousands of developers to support a custom fork of 3.2 just for fixes like these when they could have just been merged instead of going "It's fixed in 4.0, fuck em"?
> 
> The whole "if you don't like it make your own fork" attitude is unnecessarily hostile towards users and gamedevs.

Oh sorry, I thought you really needed it right now for a project of yours. I liked this PR a lot too but apparently the methods behind it may lead to a few different issues down the line. There has been a pretty extensive discussion about it in #35945.

Also, AFAIK it is planned to fix them properly in this version too, but this simply isn't the right solution.

**Comment by Riteo** [comment]
@Two-Tone sorry for the double-post, but it seems also that @pouleyKetchoupp has actually ported at least some of the fixes he's made to 3.2. #46149 is an example of this.

**Comment by vesk4000** [comment]
> And again, what about all the other people who'll be still using 3.2? Do you really expect thousands of developers to support a custom fork of 3.2 just for fixes like these when they could have just been merged instead of going "It's fixed in 4.0, fuck em"?
> 
> The whole "if you don't like it make your own fork" attitude is unnecessarily hostile towards users and gamedevs.

I mean at the end of the day at some point any team just has to push future updates to a new version of their product. They can't support older versions forever. To be honest I'm glad that they hired @pouleyKetchoupp and these physics issues are getting fixed at all in 4.0 as I wasn't sure that would be happening any time soon. Yeah maybe they should just merge this PR as a kind of quick fix before 4.0 comes out. But I don't really see much of a problem as I just compiled it myself and have been using it while waiting for 4.0. Also I don't know really know why you might want to use 3.2 instead of 4.0 unless it's hard to transition a project (which it hopefully shouldn't be), although I do respect that there might be some reasons to do so, but then again you could just compile this yourself and from my experience it works perfectly well.

**Comment by FeralBytes** [comment]
@madmiraal Having read this and the related PRs very throughly, I know that this PR will not be merged. But is it still possible that per the discussion in [#35945](https://github.com/godotengine/godot/pull/35945) that your technique could still be put to the test in the tests? So that we might understand if the Unique but still possible cases that this solution will not work, to see if they actually exist? Or have you concluded that the "U" "anisotropy limitation" scenario mentioned by Reduz would indeed not be solvable in a single iteration? Thank you for your time and for fueling fixes faster than they other wise would have landed.

**Comment by madmiraal** [comment]
@FeralBytes, the 'U' scenario presented in [#35945](https://github.com/godotengine/godot/pull/35945#issuecomment-771338129), is a [red herring](https://en.wikipedia.org/wiki/Red_herring). As [presented previously](https://github.com/godotengine/godot/pull/35945#issuecomment-768427467), '_in reality, things either fit or not but not "just fit" (I mean whathever you collide to has to actually reach this scenario from a previous one where it's not colliding, which is likely impossible) so in these cases, users just realize this and shrink the kinematic body a tiny bit to solve the problem. Just think about it, you will realize it's not a realistic use case._'

The 'U' scenario presented is not a realistic use case. No approach and no finite number of iterations will resolve this situation. Furthermore, the solution: jumping to a collision-free zone somewhere outside of the enclosed space is not a desirable outcome either; so I don't think it's a scenario worth discussing.

What can cause a problem is how Godot physics tries to avoid the possibility of unwanted collisions being detected when moving perpendicular to a surface due to floating point numerical impression. Godot physics moves the `KinematicBody` away from all the collision surfaces by the `Safe Margin`. This is done by increasing the size of the `CollisionShape`s by the `Safe Margin` during the recovery step. An issue arises when the `KinematicBody` is between two parallel collisions surfaces. When the size is increased, the `CollisionShape` can go from fitting to not fitting. The current approach provides an iterative solution to better place the `KinematicBody` between the two surfaces. The approach used in this PR may result in the `KinematicBody` being closer to the opposite surface. However, this potential issue cannot be resolved by increasing the number of iterations. Instead, as suggested previously, '_users just realize this and shrink the kinematic body a tiny bit to solve the problem._`

Note: This is not a problem in the Bullet physics solution, because, instead of moving the `KinematiBody` away from the collision surfaces by increasing the size of the `KinematicBody` by the `Safe Margin`, the `Safe Margin` is used as a minimum collision depth during the motion. A second recovery stage is performed after the motion to correct any resulting overlap. This means that the Bullet physics' solution never has a change in state from fitting to not fitting; so there is no issue.

**Comment by akien-mga** [comment]
Closing as the `master` PR was rejected, see https://github.com/godotengine/godot/pull/35945#issuecomment-909407773 and PR discussion for details.
